bpf/nodeport: split ipv4/6 nodeport nat call into egress/ingress

sahid opened the pull request:
Having the process running on two different tail calls will decrease
the verifier complexity as well as improving maintainability of the
code.

Signed-off-by: Sahid Orentino Ferdjaoui <sahid.ferdjaoui@industrialdiscipline.com>

sahid committed:
bpf/nodeport: split the cilium call ipv4 nodeport nat

Having the process running on two different tail calls will decrease
the verifier complexity as well as improving maintenability of the
code.

Signed-off-by: Sahid Orentino Ferdjaoui <sahid.ferdjaoui@industrialdiscipline.com>

brb commented on bpf/lib/nodeport.h:
```diff
@@ -1258,8 +1258,7 @@ static __always_inline int nodeport_nat_ipv4_fwd(struct __ctx_buff *ctx)
 	if (snat_v4_needed(ctx, &target.addr, &from_endpoint))
-		ret = snat_v4_process(ctx, NAT_DIR_EGRESS, &target,
-				      from_endpoint);
+		ret = snat_v4_process(ctx, NAT_DIR_EGRESS, &target, from_endpoint);
```
Nit: this change is not needed.

brb commented on bpf/lib/nodeport.h:
```diff
@@ -1607,12 +1606,58 @@ int tail_nodeport_ipv4_dsr(struct __ctx_buff *ctx)
 	return send_drop_notify_error(ctx, 0, ret, CTX_ACT_DROP, METRIC_EGRESS);
 }
-#endif /* ENABLE_DSR */
+#endif /* ENABLE_DR */
```
Nit: this change is not needed.

brb commented on bpf/lib/nodeport.h:
```diff
@@ -1836,6 +1852,7 @@ static __always_inline int nodeport_lb4(struct __ctx_buff *ctx,
 		if (nodeport_uses_dsr4(&tuple))
 			return CTX_ACT_OK;
 #endif
+		/* TODO(sahid): We could probably remove this meta. */
```
Nit: yep, please do.

brb commented on bpf/lib/nodeport.h:
```diff
@@ -1917,8 +1934,9 @@ static __always_inline int nodeport_lb4(struct __ctx_buff *ctx,
 #endif /* DSR_ENCAP_MODE */
 			ep_tail_call(ctx, CILIUM_CALL_IPV4_NODEPORT_DSR);
 		} else {
+			/* TODO(sahid): We could probably remove this meta. */
```
Ditto.

brb commented on test/bpf/check-complexity.sh:
```diff
@@ -26,27 +26,28 @@ function annotate_section_names {
-	    -e "s/\(section '2\/30'\)/\1 (tail_call IPV6_CT_INGRESS)/" \
-	    -e "s/\(section '2\/31'\)/\1 (tail_call IPV6_CT_INGRESS_POLICY_ONLY)/" \
-	    -e "s/\(section '2\/32'\)/\1 (tail_call IPV6_CT_EGRESS)/"
+	    -e "s/\(section '2\/15'\)/\1 (tail_call IPV4_NODEPORT_NAT_EGRESS)/" \
```
I've tried running (from the dev VM) the `verifier-test.sh`, and it's failing (the `master` branch works though):

```
sudo ./test/bpf/verifier-test.sh -v
...
libbpf: prog 'from_network': unrecognized ELF section name 'from-network'
libbpf: couldn't reuse pinned map at '/sys/fs/bpf/tc//globals/test_cilium_calls_65535': parameter mismatch
libbpf: map 'test_cilium_calls_65535': error reusing pinned map
libbpf: map 'test_cilium_calls_65535': failed to create: Invalid argument(-22)
libbpf: failed to load object './test/bpf/../../bpf/bpf_network.o'
Unable to load program

sahid replied:
Yes I'm actually also not able to make it passes locally. But looks like it passed with CI for `/test-1.16-4.9` Would you make a try for `k8s-1.22-kernel-5.4 (test-1.22-5.4)` ?

I have addressed all your comments but prefer to wait your back on this before force-pushing

brb replied:
I was wrong re the `verifier-test.sh` breakage - it works after I have fetched the latest https://github.com/cilium/iproute2 (alternatively, you can use either the dev VM or the cilium image to run it). However, the verifier complexity output is still missing. The instr count diff I am getting for your changes rebased against the latest master branch is the following:
```
processed 151399 insns <--- CILIUM_CALL_IPV4_NODEPORT_NAT

vs

processed 3598 insns <--- CILIUM_CALL_IPV4_NODEPORT_NAT_EGRESS
processed 80144 insns <--- CILIUM_CALL_IPV4_NODEPORT_NAT_INGRESS
```

sahid replied:
Cool thank you for that double check!

It is not so bad, no? A question, how have you been able to extract those information related to insns?

brb replied:
Actually, `processed X insns` is the verifier complexity. Yep, it's already a nice improvement! I expect both to become even smaller after the split of `snat_v4_process()`.

brb replied:
I've run the following:
```
docker run --name foobar --rm -ti --privileged -v $PWD:/cilium quay.io/cilium/cilium-ci:latest /bin/bash
mount -t bpf bpf /sys/fs/bpf
cd /cilium
make -C bpf/ KERNEL=netnext
./test/bpf/check-complexity.sh
```

brb commented on bpf/lib/common.h:
```diff
@@ -79,25 +79,26 @@
-#define CILIUM_CALL_SIZE			33
+#define CILIUM_CALL_IPV4_NODEPORT_NAT_EGRESS	15
+#define CILIUM_CALL_IPV4_NODEPORT_NAT_INGRESS	16
+#define CILIUM_CALL_IPV6_NODEPORT_NAT		17
```
We need the similar new two calls for the IPv6 too.

sahid replied:
Yes I wanted to get your approval before to work on this part. I'm considering your are good with this approach! :-)

brb reviewed (changes requested):
I did the first round of reviewing. The 4.9 CI failure might be legit (currently rerunning to be sure that's not a flake).

pchaigno commented on bpf/lib/common.h:
```diff
@@ -79,25 +79,26 @@
+#define CILIUM_CALL_IPV6_CT_INGRESS		31
+#define CILIUM_CALL_IPV6_CT_INGRESS_POLICY_ONLY	32
+#define CILIUM_CALL_IPV6_CT_EGRESS		33
+#define CILIUM_CALL_SIZE			34
```
Best to add at the end of the list to avoid changing all indexes in the tail call map.

sahid replied:
Sure I will change that.

sahid committed:
bpf/nodeport: split the cilium call ipv6 nodeport nat

Having the process running on two different tail calls will decrease
the verifier complexity as well as improving maintenability of the
code.

Signed-off-by: Sahid Orentino Ferdjaoui <sahid.ferdjaoui@industrialdiscipline.com>

brb commented on bpf/lib/common.h:
```diff
@@ -79,8 +79,8 @@
 #define CILIUM_CALL_IPV6_TO_ENDPOINT		14
-#define CILIUM_CALL_IPV4_NODEPORT_NAT		15
-#define CILIUM_CALL_IPV6_NODEPORT_NAT		16
+#define CILIUM_CALL_IPV4_NODEPORT_NAT		15  /* not used anymore. */
```
I think @pchaigno meant to use 15 for `CILIUM_CALL_IPV4_NODEPORT_NAT_EGRESS` and 16 for `CILIUM_CALL_IPV6_NODEPORT_NAT_EGRESS`, and then define the rest as 36 and 37.

sahid replied:
Oh my mistake, I fixed that in the last push.

brb commented on test/bpf/check-complexity.sh:
```diff
@@ -26,7 +26,7 @@ function annotate_section_names {
 	    -e "s/\(section '2\/14'\)/\1 (tail_call IPV6_TO_ENDPOINT)/" \
-	    -e "s/\(section '2\/15'\)/\1 (tail_call IPV4_NODEPORT_NAT)/" \
+	    -e "s/\(section '2\/15'\)/\1 (tail_call IPV4_NODEPORT_NAT_EGRESS)/" \
 	    -e "s/\(section '2\/16'\)/\1 (tail_call IPV6_NODEPORT_NAT)/" \
```
This needs to be renamed to `IPV6_NODEPORT_NAT_EGRESS`.

sahid replied:
Whaoo good catch! :-) thanks

brb commented on bpf/lib/nodeport.h:
```diff
@@ -1619,10 +1636,56 @@ int tail_nodeport_ipv4_dsr(struct __ctx_buff *ctx)
+		 * done inside a tail call here.
+		 */
+		bpf_skip_nodeport_set(ctx);
+		ep_tail_call(ctx, CILIUM_CALL_IPV4_FROM_LXC);
```
This needs to be called `CILIUM_CALL_IPV4_FROM_NETDEV`. Please rebase your branch against the latest master, it should include the change.

sahid replied:
Thanks! I would have certainly missed that after the rebase

brb reviewed (changes requested):
Getting closer to be merged, two minor comments from my side. Also, please rebase against the latest master - it should fix the CI failures too.

brb reviewed (approved):
Thanks!
